File: app/src/main/java/com/kunkunapp/allvideodowloader/browser/AdBlocker.java
package com.kunkunapp.allvideodowloader.browser;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import com.kunkunapp.allvideodowloader.R;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.net.URL;
import java.net.URLConnection;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

class AdBlocker implements Serializable {
    private List<String> filters;
    private String easylistLastModified;

    AdBlocker() {
        filters = new ArrayList<>();
    }

    public void update(final Context context) {
        final SharedPreferences prefs = context.getSharedPreferences("settings", 0);
        final String today = new SimpleDateFormat("dd MM yyyy", Locale.getDefault()).format(new Date());
        if (!today.equals(prefs.getString(context.getString(R.string.adFiltersLastUpdated), ""))) {
            new Thread() {
                @Override
                public void run() {
                    String easyList = "https://easylist.to/easylist/easylist.txt";
                    List<String> tempFilters = new ArrayList<>();
                    try {
                        URLConnection uCon = new URL(easyList).openConnection();
                        if (uCon != null) {
                            InputStream in = uCon.getInputStream();
                            BufferedReader reader = new BufferedReader(new InputStreamReader(in));
                            String line;
                            while ((line = reader.readLine()) != null) {
                                if (line.contains("Last modified")) {
                                    if (line.equals(easylistLastModified)) {
                                        reader.close();
                                        in.close();
                                        return;
                                    } else {
                                        easylistLastModified = line;
                                    }
                                } else if (!line.startsWith("!") || !line.startsWith("[")) {
                                    tempFilters.add(line);
                                }
                            }
                            if (!tempFilters.isEmpty()) {
                                filters = tempFilters;
                                Log.i("VDDebug", "updating ads filters complete. Total: " +
                                        filters.size());
                            }
                            File file = new File(context.getFilesDir(), "ad_filters.dat");
                            FileOutputStream fileOutputStream = new FileOutputStream(file);
                            try (ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileOutputStream)) {
                                objectOutputStream.writeObject(AdBlocker.this);
                            }
                            fileOutputStream.close();
                        }
                        prefs.edit().putString(context.getString(R.string.adFiltersLastUpdated), today).apply();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }.start();
        }
    }

    public boolean checkThroughFilters(String url) {
        for (String filter : filters) {
            String filterFormat = filter.replace("||", "//");
            if (url.contains(filterFormat)) {
                Log.d("VDDebug", "checkThroughFilters: " + filter + " " +  url);
                return true;
            }
        }
        return false;
    }

}

File: app/src/main/java/com/cyberIyke/allvideodowloader/browser/AdBlocker.kt
package com.cyberIyke.allvideodowloader.browser

import android.content.Context
import android.util.Log
import com.cyberIyke.allvideodowloader.R
import java.io.*
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*

internal class AdBlocker : Serializable {
    private var filters: List<String>
    private var easylistLastModified: String? = null

    init {
        filters = ArrayList()
    }

    fun update(context: Context) {
        val prefs = context.getSharedPreferences("settings", 0)
        val today = SimpleDateFormat("dd MM yyyy", Locale.getDefault()).format(Date())
        if (today != prefs.getString(context.getString(R.string.adFiltersLastUpdated), "")) {
            object : Thread() {
                override fun run() {
                    val easyList = "https://easylist.to/easylist/easylist.txt"
                    val tempFilters: MutableList<String> = ArrayList()
                    try {
                        val uCon = URL(easyList).openConnection()
                        if (uCon != null) {
                            val `in` = uCon.getInputStream()
                            val reader = BufferedReader(InputStreamReader(`in`))
                            var line: String = ""
                            while (reader.readLine()?.also { line = it } != null) {
                                if (line.contains("Last modified")) {
                                    easylistLastModified = if (line == easylistLastModified) {
                                        reader.close()
                                        `in`.close()
                                        return
                                    } else {
                                        line
                                    }
                                } else if (!line.startsWith("!") || !line.startsWith("[")) {
                                    tempFilters.add(line)
                                }
                            }
                            if (!tempFilters.isEmpty()) {
                                filters = tempFilters
                                Log.i(
                                    "VDDebug", "updating ads filters complete. Total: " +
                                            filters.size
                                )
                            }
                            val file = File(context.filesDir, "ad_filters.dat")
                            val fileOutputStream = FileOutputStream(file)
                            ObjectOutputStream(fileOutputStream).use { objectOutputStream ->
                                objectOutputStream.writeObject(
                                    this@AdBlocker
                                )
                            }
                            fileOutputStream.close()
                        }
                        prefs.edit()
                            .putString(context.getString(R.string.adFiltersLastUpdated), today)
                            .apply()
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                }
            }.start()
        }
    }

    fun checkThroughFilters(url: String): Boolean {
        for (filter in filters) {
            val filterFormat = filter.replace("||", "//")
            if (url.contains(filterFormat)) {
                Log.d("VDDebug", "checkThroughFilters: $filter $url")
                return true
            }
        }
        return false
    }
}